fw/mmap_buffer.c

```diff
@@ -0,0 +1,305 @@
+};
+
+static const struct vm_operations_struct dev_vm_ops = {
+	.close = dev_file_vm_close
+};
+
+void
+tfw_mmap_buffer_get_room(TfwMmapBufferHolder *holder,
+						 char **part1, unsigned int *size1,
+						 char **part2, unsigned int *size2)
+{
+	TfwMmapBuffer *buf = *this_cpu_ptr(holder->buf);
+	u64 head, tail;
+
+	*size2 = 0;
+
+	if (!atomic_read(&buf->is_ready)) {
+		*size1 = 0;
+		return;
+	}
+
+	head = buf->head % buf->size;
+	tail = smp_load_acquire(&buf->tail) % buf->size;
+
+	*part1 = buf->data + head;
+
+	if (head < tail) {
+		*size1 = tail - head - 1;
+		return;
+	}
+
+	if (unlikely(head == 0)) {
+		*size1 = buf->size - 1;
+	} else {
+		*size1 = buf->size - head;
+		*part2 = buf->data;
+		*size2 = tail - 1;
+	}
+}
+
+void
+tfw_mmap_buffer_commit(TfwMmapBufferHolder *holder, unsigned int size)
+{
+	TfwMmapBuffer *buf = *this_cpu_ptr(holder->buf);
+
+	smp_store_release(&buf->head, buf->head + size);
+}
+
+static int
+dev_file_open(struct inode *ino, struct file *filp)
+{
+	TfwMmapBufferHolder *holder;
+	int i;
+
+	for (i = 0; i < holders_cnt; ++i) {
+		if (strcmp(holders[i]->dev_name, (char *)filp->f_path.dentry->d_iname) == 0) {
+			holder = holders[i];
+			goto found;
+		}
+	}
+
+	return -EINVAL;
+
+found:
+	if (atomic_read(&holder->is_freeing))
+		return -ENOENT;
+	if (atomic_read(&holder->dev_is_opened))
+		return -EBUSY;
+
+	atomic_set(&holder->dev_is_opened, 1);
+	filp->private_data = holder;
+
+	return 0;
+}
+
+static int
+dev_file_close(struct inode *ino, struct file *filp)
+{
+	TfwMmapBufferHolder *holder = filp->private_data;
+
+	atomic_set(&holder->dev_is_opened, 0);
+	return 0;
+}
+
+/*
+ * This function handles the mapping of ring buffers into user space. Each
+ * buffer should be mapped by user space with an offset calculated as
+ * full_buffer_size * cpu_num, where full_buffer_size is the size of buffer data
+ * plus TFW_MMAP_BUFFER_DATA_OFFSET; cpu_num is a number of CPU in a row. This
+ * allows determining which CPU's buffer should be mapped based on the offset.
+ */
+static int
+dev_file_mmap(struct file *filp, struct vm_area_struct *vma)
+{
+	TfwMmapBufferHolder *holder = filp->private_data;
+	TfwMmapBuffer *buf, *this_buf = *this_cpu_ptr(holder->buf);
+	unsigned long pfn, size, buf_size, buf_pages;
+	int cpu_num, cpu_id;
+
+	buf_size = TFW_MMAP_BUFFER_FULL_SIZE(this_buf->size);
+	size = vma->vm_end - vma->vm_start;
+	if (size > buf_size)
+		return -EINVAL;
+
+	buf_pages = buf_size / PAGE_SIZE;
+
+#define NTH_ONLINE_CPU(n) ({ \
+	int cpu, res = -1, i = 0; \
+	for_each_online_cpu(cpu) { \
+		if (i == n) { \
+			res = cpu; \
+			break; \
+		} \
+		++i; \
+	} \
+	res; \
+})
+
+	cpu_num = vma->vm_pgoff / buf_pages;
+	cpu_id = NTH_ONLINE_CPU(cpu_num);
+	if (cpu_id < 0)
+		return -EINVAL;
+
+	buf = *per_cpu_ptr(holder->buf, cpu_id);
+	pfn = page_to_pfn(virt_to_page(buf));
+
+	if (remap_pfn_range(vma, vma->vm_start, pfn, size, vma->vm_page_prot))
+		return -EAGAIN;
+
+	vma->vm_ops = &dev_vm_ops;
+	(void)dev_vm_ops;
```

Comment:
I din't get the sense of `(void)dev_vm_ops` statement...

Reply:
Didn't notice, left after debugging.